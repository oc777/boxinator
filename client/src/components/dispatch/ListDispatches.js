import React from 'react';

export function ListDispatches() {
    return (
        <div className='listboxes'>
            <h1>List Dispatches</h1>
            <p>Here be dispatches... or dragons... or tables...</p>
        </div>
    );
}
