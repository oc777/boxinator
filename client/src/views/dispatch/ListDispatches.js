import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { fetchDispatch } from '../../store/actions/dispatchActions'

class ListDispatches extends Component {
  componentDidMount () {
    // fetch
    this.props.fetchData()
  }

  render () {
    const data = this.props.data.dispatchReducer

    let table
    let totalWeight = 0
    let totalCost = 0

    if (data.dispatch || []) {
      table = data.dispatch.map((item, index) => {
        totalWeight += item.weight
        totalCost += item.shippingCost

        const style = {
          backgroundColor: 'rgb(' + item.color + ')',
          minWidth: '90px'
        }
        return (
          <tr key={index}>
            <td>{item.nameReceiver}</td>
            <td>{item.weight} kilograms</td>
            <td style={style} />
            <td>{item.shippingCost.toFixed(2)} SEK</td>
          </tr>
        )
      })
    }

    return (
      <>
        <div className='list-boxes'>
          <h1>List Dispatches</h1>
          <p>The list of boxes, including total weight and shipping cost</p>
        </div>

        <div className='info'>
          {data.pending && <p className='loading'>Loading...</p>}
          {data.error && <p className='error'>Something went terribly wrong here :(</p>}
        </div>

        <table className='dispatch-table'>
          <thead>
            <tr>
              <th>Receiver</th>
              <th>Weight</th>
              <th>Box colour</th>
              <th>Shipping cost</th>
            </tr>
          </thead>
          <tbody>
            {table}
          </tbody>
          <tfoot>
            <tr>
              <td>Total</td>
              <td>{totalWeight.toFixed(2)} kilograms</td>
              <td />
              <td>{totalCost.toFixed(2)} SEK</td>
            </tr>
          </tfoot>

        </table>

      </>
    )
  }
}

const mapStateToProps = state => ({
  data: state
})

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchData: fetchDispatch
}, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ListDispatches)

export { ListDispatches }