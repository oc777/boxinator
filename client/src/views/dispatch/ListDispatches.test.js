import { ListDispatches } from './ListDispatches'
import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

const defaultProps = {
  data: {
    dispatchReducer: {
      dispatch:[
        { nameReceiver: "Bilbo", weight: 1.2, color: "#2a2a2a", shippingCost: 2.3 },
        { nameReceiver: "Frodo", weight: 0.45, color: "#2a2a2a", shippingCost: 8 }
      ],
      pending: false,
      error: null
    }
  },
  fetchData: jest.fn()
}

describe('ListDispatches snapshot', () => {
  it('renders without crashing with required props', () => {
    const wrapper = renderer.create(<ListDispatches {...defaultProps} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})

describe('ListDispatches default state', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(<ListDispatches {...defaultProps} />)
  })

  it('should render a table', () => {
    expect(wrapper.find('table').length).toBe(1)
    expect(wrapper.find('th').length).toBe(4)
  })

  it('should fill a table cell with color', () => {
    const row = wrapper.find('tbody').children().at(0)
    const color = row.children().at(2)
    //console.log(color.debug())
    expect(color.prop('style').backgroundColor).toMatch(/2a2a2a/)
  })

  it('should calculate total weight and cost', () => {
    const row = wrapper.find('tfoot').children().at(0)
    const weight = row.children().at(1)
    const cost = row.children().at(3)

    const totalWeight = 
      defaultProps.data.dispatchReducer.dispatch[0].weight + 
      defaultProps.data.dispatchReducer.dispatch[1].weight
    const totalCost = 
      defaultProps.data.dispatchReducer.dispatch[0].shippingCost + 
      defaultProps.data.dispatchReducer.dispatch[1].shippingCost

    expect(weight.text()).toContain(totalWeight)
    expect(cost.text()).toContain(totalCost)
  })
})
