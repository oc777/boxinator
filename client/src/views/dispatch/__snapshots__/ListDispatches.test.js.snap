// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListDispatches snapshot renders without crashing with required props 1`] = `
Array [
  <div
    className="list-boxes"
  >
    <h1>
      List Dispatches
    </h1>
    <p>
      The list of boxes, including total weight and shipping cost
    </p>
  </div>,
  <div
    className="info"
  />,
  <table
    className="dispatch-table"
  >
    <thead>
      <tr>
        <th>
          Receiver
        </th>
        <th>
          Weight
        </th>
        <th>
          Box colour
        </th>
        <th>
          Shipping cost
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Bilbo
        </td>
        <td>
          1.2
           kilograms
        </td>
        <td
          style={
            Object {
              "backgroundColor": "rgb(#2a2a2a)",
              "minWidth": "90px",
            }
          }
        />
        <td>
          2.30
           SEK
        </td>
      </tr>
      <tr>
        <td>
          Frodo
        </td>
        <td>
          0.45
           kilograms
        </td>
        <td
          style={
            Object {
              "backgroundColor": "rgb(#2a2a2a)",
              "minWidth": "90px",
            }
          }
        />
        <td>
          8.00
           SEK
        </td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <td>
          Total
        </td>
        <td>
          1.65
           kilograms
        </td>
        <td />
        <td>
          10.30
           SEK
        </td>
      </tr>
    </tfoot>
  </table>,
]
`;
